cluster_tools/tests/test_slurm.py

```diff
@@ -168,85 +169,149 @@ def wait_until_first_job_was_submitted(
+                time.sleep(0.5)
 
     finally:
         call("echo y | sacctmgr modify qos normal set MaxSubmitJobs=-1")
 
 
-def test_slurm_job_canceling_on_shutdown() -> None:
+@pytest.mark.timeout(30)
+def test_slurm_job_canceling_on_shutdown(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
     # Test that scheduled jobs are canceled on shutdown, regardless
     # of whether they are pending or running.
-    max_running_size = 2
+    monkeypatch.setenv("SLURM_MAX_RUNNING_SIZE", "2")
+    monkeypatch.setenv("SIGTERM_WAIT_IN_S", "0")
 
     executor = cluster_tools.get_executor("slurm", debug=True)
-    # Only two jobs can run at once, so that some of the jobs will be
-    # running and some will be pending.
-    original_max_running_size = os.environ.get("SLURM_MAX_RUNNING_SIZE")
-    os.environ["SLURM_MAX_RUNNING_SIZE"] = str(max_running_size)
 
-    try:
+    with executor:
+        # Only two jobs can run at once, so that some of the jobs will be
+        # running and some will be pending.
         executor.map_to_futures(sleep, [10] * 4)
 
         # Wait until first job is running
         wait_until_first_job_was_submitted(executor, "RUNNING")
 
         job_start_time = time.time()
 
-        # The job cancellation is flaky if jobs just switched from PENDING to RUNNING,
-        # probably because Python is not yet able to react to the SIGINT signal.
-        # This is an unsolved issue as of now.
-        time.sleep(1)
-
-        executor.handle_kill(signal.default_int_handler, None, None)
+        sigint_handler = signal.getsignal(signal.SIGINT)
+        assert callable(sigint_handler)  # Mainly for typechecking
+        sigint_handler(signal.SIGINT, None)
 
         # Wait for scheduled jobs to be canceled, so that the queue is empty again
         # and measure how long the cancellation takes
         while executor.get_number_of_submitted_jobs() > 0:
             time.sleep(0.5)
 
+    job_cancellation_duration = time.time() - job_start_time
+
+    # Killing the executor should have canceled all submitted jobs, regardless
+    # of whether they were running or pending in much less time than it would
+    # have taken the jobs to finish on their own
+    assert job_cancellation_duration < 5
+
+
+@pytest.mark.timeout(30)
+def test_slurm_signal_handling(
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    monkeypatch.setenv("SIGTERM_WAIT_IN_S", "0")
+    original_sigint_handler_was_called = False
+
+    def original_sigint_handler(_signum: int | None, _frame: Any) -> None:
+        nonlocal original_sigint_handler_was_called
+        original_sigint_handler_was_called = True
+
+    signal.signal(
+        signal.SIGINT,
+        original_sigint_handler,
+    )
+
+    assert len(cluster_tools.SlurmExecutor._shutdown_hooks) == 0
+    # Reset class variable after changing signal handler so it will be set again when the next cluster executor is instantiated
+    cluster_tools.SlurmExecutor._installed_signal_handler = False
```

Comment:
won't this lead to the signal handler being set up twice? how about a class method à la `uninstall_signal_handler` which would do the uninstall + set the variable to false? however, it might be overkill if it's only necessary for this test.

Reply:
In lines 262-264 we "reset" the signal handler, so the cluster executor signal handler that was installed by one of the previous tests is no longer installed. Therefore, we also need to make sure the `_installed_signal_handler` class variable is set to False again.
It's a special case for this test, because we pretend that a signal handler is already installed before the cluster executor is instantiated (to mimic what happens if voxelytics instantiates a cluster executor) and check whether it was actually called, then.